scout/ecm_prep.py

```diff
@@ -12843,110 +12938,140 @@ def prepare_packages(packages, meas_update_objs, meas_summary,
-                measure_list_package, p["name"], p["benefits"],
-                handyvars, handyfiles, opts, convert_data)
-            # Record heating/cooling equipment and envelope overlaps in
-            # package after confirming that envelope measures are present
-            if len(packaged_measure.contributing_ECMs_env) > 0:
-                packaged_measure.htcl_adj_rec(opts)
-            # Merge measures in the package object
-            packaged_measure.merge_measures(opts)
-            # Print update on measure status
-            print("Success")
-
-        # Add the new packaged measure to the measure list (if it exists)
-        # for further evaluation like any other regular measure
-        if packaged_measure is not False:
-            meas_update_objs.append(packaged_measure)
+                # Instantiate measure package object based on packaged measure
+                # subset above
+                packaged_measure = MeasurePackage(
+                    measure_list_package, p["name"], p["benefits"],
+                    handyvars, handyfiles, opts, convert_data)
+                # Record heating/cooling equipment and envelope overlaps in
+                # package after confirming that envelope measures are present
+                if len(packaged_measure.contributing_ECMs_env) > 0:
+                    packaged_measure.htcl_adj_rec(opts)
+                # Merge measures in the package object
+                packaged_measure.merge_measures(opts)
+                # Print update on measure status
+                print("Success")
+
+            # Add the new packaged measure to the measure list (if it exists)
+            # for further evaluation like any other regular measure
+            if packaged_measure is not False:
+                meas_update_objs.append(packaged_measure)
+        except Exception:
+            prep_error(p["name"], handyvars, handyfiles)
 
     return meas_update_objs
 
 
+def prep_error(meas_name, handyvars, handyfiles):
+    """Prepare and write out error messages for skipped measures/packages.
+
+    Args:
+        meas_name (str): Measure or package name.
+        handyvars (object): Global variables of use across Measure methods.
+        handyfiles (object): Input files of use across Measure methods.
+    """
+    # Complete the update to the console for each measure being processed
+    print("Error")
+    # Pull full error traceback
+    err_dets = traceback.format_exc()
+    # Construct error message to write out
+    err_msg = (
+        "\nECM '" + meas_name + "' produced the following exception that "
+        "prevented its preparation:\n" + str(err_dets) + "\n")
+    # Add ECM to skipped list
+    handyvars.skipped_ecms.append(meas_name)
+    # Print error message if in verbose mode
+    verboseprint(opts.verbose, err_msg)
+    # Write message to log file
+    with open(handyfiles.prep_err_log, "a") as ef:
+        ef.write(err_msg)
+        ef.close()
```

Comment:
It may be better to use the Python [logging](https://docs.python.org/3/library/logging.html) library instead, which is more flexible, standardized, and optimized. Maybe just as important it is thread-safe for potential multithreaded approaches that may be used in the future.

You would configure the logger up top, then just write the error with `logging.error(err_msg)`. 

An example usage is found in run_workflow.py: https://github.com/trynthink/scout/blob/master/tests/integration_testing/run_workflow.py#L16 If you want to just write errors, level would be `logging.ERROR`, or you could write all infos, warnings, and errors with `logging.INFO`. There is a `filename` arg too that can be set to prep_err_log.

Reply:
Thanks very helpful. Switching to using this module.

Reply:
Looks good to me!